difference mark

mbostock opened the pull request:
First cut at a difference mark for #159.

Fil commented on src/marks/difference.js:
```diff
@@ -0,0 +1,109 @@
+    this.positiveColor = maybeColor(positiveColor);
+    this.negativeColor = maybeColor(negativeColor);
+  }
+  filter(index) {
+    return index;
+  }
```
I don't understand why we're not using the default filter; if any point's y1 or y2 is invalid, the whole area disappears.

mbostock replied:
I copied this from area/line. I think the idea is that it uses the _line_.defined and _area_.defined check instead to interrupt the path, rather pretending the point doesn’t exist and connecting across the gap. I’ll add some tests to make sure this works.

mbostock replied:
The filtering is handled by groupIndex, which I’ve now restored.

Fil commented on src/marks/difference.js:
```diff
@@ -0,0 +1,109 @@
+function renderArea(X, Y, y0, {curve}, index) {
+  return shapeArea()
+    .curve(curve)
+    .defined((i) => i >= 0)
```
I don't understand this either :) And it's not tested.

Fil commented on src/marks/difference.js:
```diff
@@ -0,0 +1,109 @@
+function renderLine(X, Y, {curve}, index) {
+  return shapeLine()
+    .curve(curve)
+    .defined((i) => i >= 0)
```
ditto

Fil commented on test/plots/difference.ts:
```diff
@@ -0,0 +1,13 @@
+  const aapl = await d3.csv<any>("data/aapl.csv", d3.autoType);
+  const goog = await d3.csv<any>("data/goog.csv", d3.autoType);
+  const x = aapl.map((d) => d.Date);
+  const y1 = aapl.map((d, i, data) => d.Close / data[0].Close);
+  const y2 = goog.map((d, i, data) => d.Close / data[0].Close);
```
It makes sense to normalize, but the chart is difficult to read: which part is "goog > aapl" and vice-versa? Maybe we can solve it by adding annotations, rather than a color legend?

mbostock replied:
I also want to try comparing week-over-week or month-over-month change. This might involve shifting _x1_ or _x2_ instead of separating _y1_ and _y2_. We should test this use case.

Fil commented on src/marks/difference.js:
```diff
@@ -0,0 +1,109 @@
+
+class Difference extends Mark {
+  constructor(data, options = {}) {
+    const {x1, y1, x2, y2, curve, tension, positiveColor = "green", negativeColor = "blue"} = options;
```
these default colors are not very pretty; maybe use blue/orange instead?

mbostock replied:
I’ll pick some prettier colors.

Fil reviewed (commented):
Nice! A few comments for now.

mbostock commented:
Now improved! I have an open question as to whether this should support the **z** channel, and what to do if **fill**, **fillOpacity**, **stroke**, **strokeOpacity**, or **strokeWidth** are specified as channels. I’m inclined to say that these are not supported (at least for now — I’m trying to keep things simple). But I also don’t protect against them, and I’m not entirely sure how we would protect against them. What do you think? Perhaps this elucidates why I think requests like #175 would be so difficult to support!

Fil commented:
Toyed with this in #1897 in a composite mark; I also worked on the default tip contents, to show y1, y2, and the actual difference = y1-y2 (which is the point of this mark).

I'd like to suggest two changes to the options, for a bit more consistency:
- use **positive** and **negative**, or **positiveFill** and **negativeFill** (instead of positiveColor and negativeColor)
- make **positiveOpacity** and **negativeOpacity** default to fillOpacity (instead of opacity)

this way **opacity** would be the usual overall mark opacity (line + areas), **strokeOpacity** would be used on the line.

Fil commented:
For the record, this (correctly) ignores the computed difference value from the default tip which I suggested in #1897. It was in fact a wrong good idea, because the difference mark is showing the _visual_ difference between scaled values, which is not necessarily encoding y2 &minus; y1. (Adding the difference to the default tip would have been correct only if we could assume a linear scale for *y*.)

Users can add it explicitly with the channels option, e.g.:
> `channels: {difference: y1.map((d, i) => d - y2[i])},`

or `d/y2[i]` if using a log scale and comparing ratios.

mbostock commented on src/marks/difference.js:
```diff
@@ -15,11 +15,11 @@ export function differenceY(
-    opacity = 1,
-    positiveOpacity = opacity,
-    negativeOpacity = opacity,
+    positive = "#01ab63",
+    negative = "#4269d0",
```
These sound like booleans to me. I definitely prefer either the *Color* or *Fill* suffix.

Fil replied:
OK, I'd favor positiveFill/negativeFill in that case (pushing this for now).

mbostock replied:
I think we should use positiveFillOpacity & negativeFillOpacity, too (for symmetry, and to make clear that these affect the fillOpacity and not the opacity of the constituent areas).

Fil commented:
Two more suggestions:

a) flip the meaning of y1 and y2, to make it closer to areaY. Instead of having y1 be the primary metric, we'd have y1 as the base metric ("bottom" when positive), and y2 as the reference metric ("top" when positive). Switching from areaY to differenceY would not mean exchanging y1 and y2.

b) refer to #1866 when it comes to drawing the line on the area.

With these two changes, differenceY would behave like areaY, except the area part would be split into its positive and negative sections. (We could even decide to make this a simple option of areaY and avoid introducing a new symbol.)

mbostock commented:
The difference mark is hard to use when you have a tidy dataset like this, of temperature observations from two stations (San Francisco and San Jose):

```
date,station,tmax,tmin
2020-12-31,SJ,59,43
2020-12-31,SF,60,47
2020-12-30,SJ,58,33
2020-12-30,SF,57,40
2020-12-29,SJ,62,38
2020-12-29,SF,61,41
2020-12-28,SJ,57,43
2020-12-28,SF,56,47
```

You can say this, but it’s pretty magical and assumes that the two stations exactly alternate rows!

```js
Plot.differenceY(temperature, Plot.windowY(14, {
  filter: (d) => d.station === "SF",
  x: "date",
  y: "tmin",
  y1: (d, i) => temperature[i + 1]?.tmin
}))
```

Perhaps we need a [pivot wider](https://tidyr.tidyverse.org/reference/pivot_wider.html) transform for cases like this. This could produce a more natural shape for this case (in tidyspeak, names_from = station, values_from = c(tmin, tmax), and names_sep = ':'… and I guess because the date column isn’t mentioned, it becomes id_cols?):

```
date,SJ:tmax,SJ:tmin,SF:tmax,SF:tmin
2020-12-31,59,43,60,47
2020-12-30,58,33,57,40
2020-12-29,62,38,61,41
2020-12-28,57,43,56,47
```

Fil commented:
Addendum to my previous remark: "positive" and "negative" get flipped when we use the `y: {reverse: true}` option. This felt a bit unexpected—in the sense that when flipping the y axis I expected the chart to flip, but instead it changed the meaning of the colors.

| usual y | flipped y |
| --- | -- |
| ![difference](https://github.com/observablehq/plot/assets/7001/41f09e55-fba7-4b96-a1ed-0e21342a3912) | ![difference-flipped](https://github.com/observablehq/plot/assets/7001/cf722b8f-fe3b-48ad-a4f2-0467e4333fa0) |
| ![negative](https://github.com/observablehq/plot/assets/7001/034ca5b6-d234-4795-8ea4-a9f119a3b3e5) | ![negative-flipped](https://github.com/observablehq/plot/assets/7001/cb6fe7bf-2159-4b7d-b612-cda9f731d650) |

Not sure if this can be fixed. At least it's a reminder that this mark is not about the (sign of the) difference, but rather about the orientation of the difference between the scaled values (i.e., up or down).

mbostock commented:
We could use the same `inferScaleOrder` technique that the axis label uses to infer the arrow orientation to address this. I think we should fix it — we want the color to represent the sign of the difference in the abstract value, not the encoded value.

mbostock commented:
I broke out #1915 and #1916 from this PR.

mbostock commented:
In the latest commits, I introduced an explicit shift transform rather than limiting it a **shift** option on the difference mark. I’ve also changed to setting a domain hint on the channel rather than filtering the data to produce the same effect; I think this is cleaner and more precise.

If you can review this, @Fil, and the related PRs that I extracted, I can finish up the documentation and we should be able to do a release this week. I’d like to move on from further polishing of this feature (and start using it in dashboards)!

Fil commented:
The **z** branch (#1906) now supports clip: "frame".

mbostock commented:
LGTY @Fil?

Fil reviewed (approved):
yes! Thanks for fixing the documentation on **z**
